apps/test_ogrsf.cpp

```diff
@@ -401,6 +401,40 @@ static int TestDataset( GDALDriver** ppoDriver )
     poDS->GetMetadata( nullptr );
     poDS->GetMetadataItem( "", nullptr );
 
+    if ( !poDriver->GetMetadataItem( GDAL_DCAP_VECTOR ))
+    {
+        printf("FAILURE: Driver does not advertize GDAL_DCAP_VECTOR!\n" );
+        GDALClose(poDS);
+        return FALSE;
+    }
+
+    // Test consistency of datasource capabilities and driver metadata
+    if ( poDS->TestCapability(ODsCCreateLayer) && !poDriver->GetMetadataItem( GDAL_DCAP_CREATE_LAYER ) )
+    {
+        printf("FAILURE: Dataset advertizes ODsCCreateLayer capability but driver metadata does not advertize GDAL_DCAP_CREATE_LAYER!\n" );
+        GDALClose(poDS);
+        return FALSE;
+    }
+    if ( !bReadOnly && poDriver->GetMetadataItem( GDAL_DCAP_CREATE_LAYER ) && !poDS->TestCapability(ODsCCreateLayer) )
+    {
+        printf("FAILURE: Driver advertizes GDAL_DCAP_CREATE_LAYER capability but dataset does not advertize ODsCCreateLayer!\n" );
```

Comment:
I can imagine a number of drivers where that would fail: those who can support only 1 layer. That should be combined with a test of GDAL_DCAP_MULTIPLE_VECTOR_LAYERS